pkg/volume/util/atomic_writer.go

```diff
@@ -0,0 +1,305 @@
+//       <target-dir>/k8s_space/annotations -> ../..<identifier-token>/k8s_space/annotations
+//       <target-dir>/..<identifier-token>  -> ..<identifier-token>.12345678
+//
+//     where ..<identifier-token>.12345678 is a randomly generated directory that
+//     contains the real data.  If a file has to be dumped in subdirectory
+//     (for example <target-dir>/user_space/labels), the writer builds a
+//     relative symlink (<target-dir>/user_space/labels -> ../..<identifier-token>/user_space/labels)
+// 5.  The previous timestamped directory is detected reading the '..<identifier-token>' symlink
+// 6.  A new temporary symlink is created: ..<identifier-token>_tmp
+// 7.  ..<identifier-token>_tmp is renamed to ..<identifier-token>
+// 8.  The previous timestamped directory is removed, if it exists
+// 9.  The sentinel file modification and access times are updated (file is created if it does not
+//     already exist)
+func (w *AtomicWriter) Write(payload map[string][]byte) error {
+	shouldWritePayload, err := w.shouldWritePayload(payload)
+	if err != nil {
+		glog.Errorf("%v: unable to determine whether payload should be written to disk: %v", w.LogContext, err)
+		return err
+	}
+
+	if !shouldWritePayload {
+		glog.V(5).Infof("%v: No update required for target directory %v", w.LogContext, w.TargetDir)
+		return nil
+	}
+
+	tsDir, err := w.newTimestampDir()
+	if err != nil {
+		glog.V(5).Infof("%v: unable to create new timestamped directory: %v", w.LogContext, err)
+		return err
+	}
+
+	err = w.writePayloadToDir(payload, tsDir)
+	if err != nil {
+		glog.Errorf("%v: unable to write payload to dir %v: %v", w.LogContext, tsDir, err)
+		return err
+	}
+
+	err = w.createPayloadSymlinksInTargetDir(payload)
+	if err != nil {
+		glog.Errorf("%v: unable to create payload symlinks in %v: %v", w.LogContext, w.TargetDir, err)
+		return err
+	}
+
+	_, tsDirName := filepath.Split(tsDir)
+	currentStateDirPath := path.Join(w.TargetDir, w.currentStateDirName())
+	oldTsDir, err := os.Readlink(currentStateDirPath)
+	if err != nil && !os.IsNotExist(err) {
+		glog.Errorf("%v: unable to read link for current state directory: %v", w.LogContext, err)
+		return err
+	}
+
+	updateStateDirPath := path.Join(w.TargetDir, w.updateStateDirName())
+	err = os.Symlink(tsDirName, updateStateDirPath)
+	if err != nil {
+		os.RemoveAll(tsDir)
+		glog.Errorf("%v: unable to create symolic link: %s", w.LogContext, err.Error())
+		return err
+	}
+
+	err = os.Rename(updateStateDirPath, currentStateDirPath)
+	if err != nil {
+		os.Remove(updateStateDirPath)
+		os.RemoveAll(tsDir)
+		glog.Errorf("%v: unable to rename symbolic link %v: %v", w.LogContext, updateStateDirPath, err)
+		return err
+	}
+	if len(oldTsDir) > 0 {
+		err = os.RemoveAll(path.Join(w.TargetDir, oldTsDir))
+		if err != nil {
+			glog.Errorf("%v: unable to remove directory %v: %v", w.LogContext, oldTsDir, err)
+		}
+	}
+
+	// TODO: determine error handling around sentinel file
+	return w.touchSentinelFile()
+}
```

Comment:
I guess here is where you should unlink files that exist but are not in the payload?  and directories if they are empty.